lib/services/elasticsearch.js

```diff
@@ -153,519 +166,479 @@ class ElasticSearch extends Service {
+        };
 
-        return this.client.index(esRequest)
-          .then(result => this.refreshIndexIfNeeded(
-            esRequest,
-            _.extend(result, {_source: request.input.body}))
-          )
-          .catch(error => Bluebird.reject(
-            this.esWrapper.formatESError(error))
-          );
-      });
+        return apiResponse;
+      })
+      .catch(error => this.esWrapper.reject(error));
   }
 
   /**
    * Sends the partial document to elasticsearch
    * with the id to update
    *
-   * @param {Request} request
-   * @returns {Promise} resolve an object that contains _id
+   * @param {string} index - Index name
+   * @param {string} collection - Collection name
+   * @param {string} id - Document id
+   * @param {object} content - Updated content
+   * @param {object} options - refresh (null), userId (null), retryOnConflict (0)
+   *
+   * @returns {Promise} { _id, _version }
    */
-  update(request) {
-    const esRequest = initESRequest(request, ['refresh', 'retryOnConflict']);
+  update (
+    index,
+    collection,
+    id,
+    content,
+    { refresh=null, userId=null, retryOnConflict=-1 })
+  {
+    const esRequest = {
+      index: this._getESIndex(index, collection),
+      id,
+      refresh,
+      retryOnConflict
+    };
 
-    esRequest.id = request.input.resource._id;
-    esRequest.body = request.input.body;
+    // injecting retryOnConflict default configuration
+    if (esRequest.retryOnConflict === 0
+      && this.config.defaults.onUpdateConflictRetries > 0
+    ) {
+      esRequest.retryOnConflict = this.config.defaults.onUpdateConflictRetries;
+    }
 
     assertNoRouting(esRequest);
     assertWellFormedRefresh(esRequest);
 
-    return this.kuzzle.indexCache.exists(esRequest.index, esRequest.type)
+    return this.kuzzle.indexCache.exists(index, collection)
       .then(exists => {
         if (! exists) {
           this.throw(
             'index_or_collection_does_not_exist',
-            esRequest.index,
-            esRequest.type
-          );
-        }
-
-        // injecting retryOnConflict default configuration
-        if (!_.has(esRequest, 'retryOnConflict')
-          && this.config.defaults.onUpdateConflictRetries > 0
-        ) {
-          esRequest.retryOnConflict = this.config.defaults.onUpdateConflictRetries;
+            index,
+            collection);
         }
 
         // Add metadata
-        esRequest.body._kuzzle_info = {
-          active: true,
+        content._kuzzle_info = {
           updatedAt: Date.now(),
-          updater: this._getUserId(request)
+          updater: String(userId)
         };
 
-        esRequest.body = {doc: esRequest.body};
+        esRequest.body = { doc: content };
 
         return this.client.update(esRequest);
       })
-      .then(result => this.refreshIndexIfNeeded(esRequest, result))
-      .catch(error => Bluebird.reject(this.esWrapper.formatESError(error)));
+      .then(({ body }) => {
+        const apiResponse = {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Immediately return this expression instead of assigning it to the temporary variable "apiResponse". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1488)